runner/src/runner.rs

```diff
@@ -0,0 +1,628 @@
+                    // To ensure there can't be two vault processes using the same Bitcoin wallet.
+                    runner.terminate_proc_and_wait()?;
+
+                    // Delete old release
+                    runner.delete_downloaded_release()?;
+
+                    // Download new release
+                    runner.download_binary(new_release).await?;
+
+                    // Run the downloaded release
+                    runner.run_binary().await?;
+                }
+            }
+            tokio::time::sleep(BLOCK_TIME).await;
+        }
+    }
+}
+
+#[async_trait]
+pub trait RunnerExt {
+    fn parachain_rpc(&self) -> &WebsocketClient;
+    fn vault_args(&self) -> &Vec<String>;
+    fn child_proc(&mut self) -> &mut Option<Child>;
+    fn set_child_proc(&mut self, child_proc: Child);
+    fn downloaded_release(&self) -> &Option<DownloadedRelease>;
+    fn set_downloaded_release(&mut self, downloaded_release: DownloadedRelease);
+    fn download_path(&self) -> &PathBuf;
+    fn set_download_path(&mut self, download_path: PathBuf);
+    async fn try_get_release(&self, pending: bool) -> Result<Option<ClientRelease>, Error>;
+    async fn download_binary(&mut self, release: ClientRelease) -> Result<(), Error>;
+    fn uri_to_bin_path(&self, uri: &str) -> Result<(String, PathBuf), Error>;
+    fn delete_downloaded_release(&mut self) -> Result<(), Error>;
+    async fn run_binary(&mut self) -> Result<(), Error>;
+    fn terminate_proc_and_wait(&mut self) -> Result<(), Error>;
+    async fn get_request_bytes(&self, url: String) -> Result<Bytes, Error>;
+}
+
+#[async_trait]
+impl RunnerExt for Runner {
+    fn parachain_rpc(&self) -> &WebsocketClient {
+        &self.parachain_rpc
+    }
+
+    fn vault_args(&self) -> &Vec<String> {
+        &self.vault_args
+    }
+
+    fn child_proc(&mut self) -> &mut Option<Child> {
+        &mut self.child_proc
+    }
+
+    fn set_child_proc(&mut self, child_proc: Child) {
+        self.child_proc = Some(child_proc);
+    }
+
+    fn downloaded_release(&self) -> &Option<DownloadedRelease> {
+        &self.downloaded_release
+    }
+
+    fn set_downloaded_release(&mut self, downloaded_release: DownloadedRelease) {
+        self.downloaded_release = Some(downloaded_release);
+    }
+
+    fn download_path(&self) -> &PathBuf {
+        &self.download_path
+    }
+
+    fn set_download_path(&mut self, download_path: PathBuf) {
+        self.download_path = download_path;
+    }
+
+    async fn try_get_release(&self, pending: bool) -> Result<Option<ClientRelease>, Error> {
+        try_get_release(self, pending).await
+    }
+
+    async fn run_binary(&mut self) -> Result<(), Error> {
+        let child = run_binary(self, Stdio::inherit()).await?;
+        self.set_child_proc(child);
+        Ok(())
+    }
+
+    async fn download_binary(&mut self, release: ClientRelease) -> Result<(), Error> {
+        let downloaded_release = download_binary(&*self, release).await?;
```

Comment:
Why not inline this function?

Reply:
So I can test unit `download_binary` by mocking all the methods in `RunnerExt` that it depends on: https://github.com/savudani8/interbtc-clients/blob/4f8acc8a8daa7a1e5722d9662bc4d4fec4e7a98f/runner/src/runner.rs#L448

This is the cleanest pattern I could come up with, let me know if there's a better approach.